waveform.py

```diff
@@ -260,35 +271,51 @@ def state_to_int32(self, state: [int]):
 
-    def wave_split(self, flip: bool = True) -> [Waveform]:
+    def wave_split(self, flip: bool = True) -> List[HSDIOWaveform]:
         """
-        splits a waveform object into a list of waveform objects with len() = dev. where dev is the number of hsdio
-        devices receiving waveforms.
+        splits a waveform object into a list of waveform objects with len() = dev. where dev is the
+        number of hsdio devices receiving waveforms.
 
-        The hsdioSession methods assume waveform data that is encoded for a maximum of 32 channels per device. The
-        waveform information passed into the hsdio class via xml encodes data for all channels which will be operated
-        on multiple devices. This function is intended to perform the necessary split of data by copying most of the
-        Waveform object info into separate objects, with the states info split between them
+        The hsdioSession methods assume waveform data that is encoded for a maximum of 32 channels
+        per device. The waveform information passed into the hsdio class via xml encodes data for
+        all channels which will be operated on multiple devices. This function is intended to
+        perform the necessary split of data by copying most of the Waveform object info into
+        separate objects, with the states info split between them
 
         Args:
             flip : should the order of the channels in each new waveform be flipped?
 
         Returns:
-            numpy array of split up waveform objects
+            list of split up waveform objects
         """
 
         dev = int(len(self.states[0])/32)
 
         # mapping may be confused in practical order of devices, maybe flip comes before split?
-        wave_array = np.empty(dev, dtype=type(self))
+        wave_array = []
         for d in range(dev):
             if flip:
-                new_states = np.array([np.flip(state[d*32:(d+1)*32])
-                                    for state in self.states])
+                new_states = np.array([np.flip(state[d*32:(d+1)*32]) for state in self.states])
             else:
-                new_states = np.array([state[d*32:(d+1)*32]
-                                    for state in self.states])
-            wave_array[d] = Waveform(self.name, self.transitions, new_states, self.data_format)
+                new_states = np.array([state[d*32:(d+1)*32] for state in self.states])
+            wave_array.append(
+                HSDIOWaveform(
+                    self.name,
+                    self.transitions,
+                    new_states,
+                    self.data_format
+                )
+            )
 
         return wave_array
-    
+
+    def check_state_len(self):
+        """
+        checks that the states array
+        Returns:
+
+        """
+        cl_str = str(self.__class__.__name__)
+        state_len = self.states.shape[0]
+        as_ms = f"{cl_str}.states.shape[0] = {state_len}; it's not divisible by 32!"
```

Comment:
the assert method here isn't specific enough to be helpful imo. something to the effect of "expected channels per card to be 32" or something similar